### PR TITLE
Add docfx language-server client

### PR DESCRIPTION
**Pull request by shanluo911**
[AB#346671](https://dev.azure.com/ceapex/d3d54af3-265a-4f18-95f6-9a46397ca583/_workitems/edit/346671)
Description:
1. Set up language client which will starts language server.
2. Add `start language server` command

Test:
1. Unit test
2. Manual test, language server started and didOpen/ didChange message received from language server.

### CODE REVIEW

File: test/utils/faker.ts [hunk 1]
```diff
@@ -147,6 +150,11 @@ export function getFakedBuildExecutor(docfxExecutionResult: DocfxExecutionResult
         cancelBuild: (): Promise<void> => {
             buildCancelled = true;
             return Promise.resolve();
+        },
+        getLanguageClient: (input: BuildInput, buildUserToken: string): LanguageClient => {
```
Comment (928PJY) [inline test/utils/faker.ts]: just return a mock client like:
```
return <LanguageClient>{
                start: () => new Disposable(() => { }),
            };s
```

File: package.json [hunk 2]
```diff
@@ -98,6 +103,11 @@
         "command": "docs.openInstallationDirectory",
         "title": "Open extension install directory",
         "category": "Docs"
+      },
+      {
+        "command": "docs.startServer",
+        "title": "Start docfx language server",
```
Comment (928PJY) [inline package.json]: how about we call it `Enable Docs real-time validation`, `language server` is too technical for the end user.

File: package.json [hunk 3]
```diff
@@ -66,6 +66,11 @@
           ],
           "default": "",
           "description": "Microsoft employee or public contributor"
+        },
+        "docs-build.realTimeValidation.enable": {
+          "type": "boolean",
+          "default": false,
+          "description": "Enable single-file real-time validation"
```
Comment (928PJY) [inline package.json]: how about we call it `Automatically enable Docs real-time validation`.

File: src/build/buildController.ts [hunk 4]
```diff
@@ -31,6 +34,10 @@ export class BuildController {
         this._instanceAvailable = true;
     }
 
+    dispose() {
+        this._dispose.dispose();
```
Comment (928PJY) [inline src/build/buildController.ts]: _dispose can be null

File: src/common/eventType.ts [hunk 5]
```diff
@@ -54,6 +54,10 @@ export enum EventType {
     TriggerCommandWithUnkownUserType,
     ExtensionActivated,
 
+    // Language Server
+    StartServerFailed,
+    LanguageClientConfigProgress,
```
Comment (928PJY) [inline src/common/eventType.ts]: reuse the `BuildProgress` and a message string property is enough?

File: src/build/buildController.ts [hunk 6]
```diff
@@ -110,17 +131,25 @@ export class BuildController {
...
 
-    private async validateUserCredential(credential: Credential) {
+    private async validateUserCredential(credential: Credential, actionType: ActionType) {
         if (this._environmentController.userType === UserType.PublicContributor) {
             return;
         }
         if (credential.signInStatus !== 'SignedIn' && this._environmentController.userType === UserType.MicrosoftEmployee) {
-            throw new DocsError(`Microsoft employees must sign in before validating.`, ErrorCode.TriggerBuildBeforeSignIn);
+            if (actionType === 'Build') {
+                throw new DocsError(`Microsoft employees must sign in before validating.`, ErrorCode.TriggerBuildBeforeSignIn);
```
Comment (928PJY) [inline src/build/buildController.ts]: use a general code like `TriggerCommandWithoutSignIn` to handle those two cases? same as below

File: src/build/buildController.ts [hunk 7]
```diff
@@ -85,6 +92,20 @@ export class BuildController {
         }
     }
 
+    public async startDocfxLanguageServer(credential: Credential): Promise<void> {
+        let buildInput: BuildInput;
+
+        try {
+            await this.validateUserCredential(credential, 'StartServer');
+            buildInput = await this.getBuildInput(credential);
+            this._client = this._buildExecutor.getLanguageClient(buildInput, credential.userInfo?.userToken);
```
Comment (928PJY) [inline src/build/buildController.ts]: any particular reason that you return a client from the executor and start it from here, instead of directly start it inside the executor?

File: src/build/buildExecutor.ts [hunk 8]
```diff
@@ -65,6 +72,43 @@ export class BuildExecutor {
...
+                command,
+                args,
+                options
+            },
+        };
+
+        let clientOptions: LanguageClientOptions = {
+            progressOnInitialization: true,
+            synchronize: {
+                configurationSection: "docfxLanguageServer",
```
Comment (928PJY) [inline src/build/buildExecutor.ts]: where do you define this configuration?

File: src/build/buildExecutor.ts [hunk 9]
```diff
@@ -65,6 +72,43 @@ export class BuildExecutor {
...
+            },
+            debug: {
+                command,
+                args,
+                options
+            },
+        };
+
+        let clientOptions: LanguageClientOptions = {
+            progressOnInitialization: true,
```
Comment (928PJY) [inline src/build/buildExecutor.ts]: what does this use for? do we need it for now?

File: src/common/docsEnvironmentController.ts [hunk 10]
```diff
@@ -60,6 +64,10 @@ export class DocsEnvironmentController implements EnvironmentController, vscode.
         return this._userType;
     }
 
+    public get enableRealTimeValidation(): boolean {
```
Comment (928PJY) [inline src/common/docsEnvironmentController.ts]: Seems we haven't used this for now, how about moving it to another PR? and all the related changes

File: src/build/buildExecutor.ts [hunk 11]
```diff
@@ -65,6 +72,43 @@ export class BuildExecutor {
...
+            synchronize: {
+                configurationSection: "docfxLanguageServer",
+                fileEvents: workspace.createFileSystemWatcher("**/*.cs"),
```
Comment (928PJY) [inline src/build/buildExecutor.ts]: we haven't use this setting yet, remove it and add back when we need it

File: src/build/buildExecutor.ts [hunk 12]
```diff
@@ -65,6 +72,43 @@ export class BuildExecutor {
...
+            },
+        };
+
+        this._eventStream.post(new LanguageClientConfigProgress(command, args, options));
+        const client = new LanguageClient("docfxLanguageServer", "Docfx Language Server", serverOptions, clientOptions);
+        client.registerProposedFeatures();
+        client.trace = Trace.Verbose;
```
Comment (928PJY) [inline src/build/buildExecutor.ts]: why we set the trace level to verbose by default

File: src/build/buildExecutor.ts [hunk 13]
```diff
@@ -65,6 +72,43 @@ export class BuildExecutor {
         return buildResult;
     }
 
+    public getLanguageClient(input: BuildInput, buildUserToken: string): LanguageClient {
+        let buildParameters = this.getBuildParameters(undefined, input, buildUserToken);
+        if (buildUserToken) {
+            buildParameters.envs["X-OP-BuildUserToken"] = buildUserToken;
+        }
+        let isPublic = !buildUserToken;
```
Comment (928PJY) [inline src/build/buildExecutor.ts]: when we already have a config to indicate the user role, we should not use this anymore, and there are several corner cases you may need to handle when we we support the user role:
1. If the user role is public contributor, you need to clean the credential stored in the local storage, or the statubar will still display the original credential information.
You can do those in s separated PR.

File: src/build/buildExecutor.ts [hunk 14]
```diff
@@ -193,4 +237,11 @@ export class BuildExecutor {
         }
         return cmdWithParameters;
     }
+
+    private getServeExecArgs(input: BuildInput, isPublicUser: boolean): string[] {
```
Comment (928PJY) [inline src/build/buildExecutor.ts]: Can we add this to the `getBuildParameters` since is also a part of build parameters

File: src/build/buildExecutor.ts [hunk 15]
```diff
@@ -65,6 +72,43 @@ export class BuildExecutor {
         return buildResult;
     }
 
+    public getLanguageClient(input: BuildInput, buildUserToken: string): LanguageClient {
+        let buildParameters = this.getBuildParameters(undefined, input, buildUserToken);
+        if (buildUserToken) {
+            buildParameters.envs["X-OP-BuildUserToken"] = buildUserToken;
```
Comment (928PJY) [inline src/build/buildExecutor.ts]: Use the name `DOCS_OPS_TOKEN` as the design spec

File: package.json (hunk 3 above)
Comment (shanluo911) [reply package.json]: move this config to next pr

File: src/build/buildExecutor.ts (hunk 8 above)
Comment (shanluo911) [reply src/build/buildExecutor.ts]: removed

File: src/common/docsEnvironmentController.ts (hunk 10 above)
Comment (shanluo911) [reply src/common/docsEnvironmentController.ts]: removed.

File: package.json [hunk 16]
```diff
@@ -98,6 +98,11 @@
         "command": "docs.openInstallationDirectory",
         "title": "Open extension install directory",
         "category": "Docs"
+      },
+      {
+        "command": "docs.startServer",
```
Comment (928PJY) [inline package.json]: make command name consistent with title

File: src/build/buildExecutor.ts [hunk 17]
```diff
@@ -194,4 +238,11 @@ export class BuildExecutor {
         }
         return cmdWithParameters;
     }
+
+    private getServeExecArgs(input: BuildInput, isPublicUser: boolean): string[] {
+        let cmdWithParameters = 'serve --language-server';
```
Comment (928PJY) [inline src/build/buildExecutor.ts]: add workingDirectory and avoid duplicate code

File: test/unitTests/build/buildController.test.ts [hunk 18]
```diff
@@ -491,4 +491,46 @@ describe('BuildController', () => {
...
+
+    it('Start language server when credential expires', async () => {
+        let opBuildAPIClient = <OPBuildAPIClient>{
+            validateCredential: (opBuildUserToken: string, eventStream: EventStream): Promise<boolean> => {
+                return new Promise((resolve, reject) => {
+                    resolve(false);
+                });
+            }
+        };
+        buildController = new BuildController(getFakedBuildExecutor(DocfxExecutionResult.Succeeded), opBuildAPIClient, undefined, fakeEnvironmentController, eventStream);
```
Comment (928PJY) [inline test/unitTests/build/buildController.test.ts:516]: Verify the `startLanguageServer` get called with the right parameters.

File: src/build/buildExecutor.ts [hunk 19]
```diff
@@ -66,6 +80,35 @@ export class BuildExecutor {
         return buildResult;
     }
 
+    public startLanguageServer(input: BuildInput, buildUserToken: string) {
+        let buildParameters = this.getBuildParameters(undefined, input, buildUserToken);
+        if (buildUserToken) {
```
Comment (928PJY) [inline src/build/buildExecutor.ts]: only set the token variable when it is a Microsoft employee?

File: src/build/buildExecutor.ts [hunk 20]
```diff
@@ -66,6 +80,35 @@ export class BuildExecutor {
...
+            debug: {
+                command,
+                args,
+                options
+            },
+        };
+
+        let clientOptions: LanguageClientOptions = {};
+
+        this._eventStream.post(new BuildProgress(`command: ${command}, args: ${args.join(' ')}`));
```
Comment (928PJY) [inline src/build/buildExecutor.ts]: how about make this more readable like the `build` since this will display on the channel?

File: src/build/buildExecutor.ts [hunk 21]
```diff
@@ -66,6 +80,35 @@ export class BuildExecutor {
...
     }
 
+    public startLanguageServer(input: BuildInput, buildUserToken: string) {
+        let buildParameters = this.getBuildParameters(undefined, input, buildUserToken);
+        if (buildUserToken) {
+            buildParameters.envs['DOCS_OPS_TOKEN'] = buildUserToken;
+        }
+        let command = this._binary;
+        let args = buildParameters.serveCommand;
+        let options = { env: buildParameters.envs, cwd: this._cwd };
```
Comment (928PJY) [inline src/build/buildExecutor.ts:90]: merge with the process environment to inherit the current process environment (reference the function `executeDocfx`)

File: test/unitTests/build/buildExecutor.test.ts [hunk 22]
```diff
@@ -368,4 +369,39 @@ describe('BuildExecutor', () => {
...
+            sinon.stub(LanguageClient.prototype, 'registerProposedFeatures').callsFake(() => {
+                return;
+            });
+            sinon.stub(LanguageClient.prototype, 'start').callsFake(() => {
+                return undefined;
+            });
+        });
+
+        it('Public contributor', async () => {
+            sinon.stub(fakedEnvironmentController, "userType").get(function getUserType() {
```
Comment (928PJY) [inline test/unitTests/build/buildExecutor.test.ts:392]: same as buildController test, you need to verify that the function `registerProposedFeatures ` and `start` get called.

File: src/build/buildExecutor.ts [hunk 23]
```diff
@@ -66,6 +80,42 @@ export class BuildExecutor {
...
+            debug: {
+                command,
+                args,
+                options: optionsWithFullEnvironment
+            },
+        };
+
+        let clientOptions: LanguageClientOptions = {};
+
+        this._eventStream.post(new BuildProgress(`Starting language server using command: ${command} ${args.join(' ')}`));
```
Comment (928PJY) [inline src/build/buildExecutor.ts:113]: minor： use the `buildParameters.serveCommand` directly to avoid unnessesary join

**Review by 928PJY — APPROVED** [review]
